Réduit la quantité de requêtes réalisées sur la page d'accueil des forums

artragis opened the pull request:
fix #3269

# But de la PR

Le but est d'améliorer la vitesse de réponse de la page d'accueil des forums. Celle ci est longue à cause du nombre ahurissant de requêtes par rapport au peu d'informations affichées.

J'ai dû distinguer deux cas : authentifié et anonyme. En effet le second permet pas mal de simplifications.

Cela peut se voir sur le benchmark rapide que j'ai fait : 
## Anonymes

![25% de réduction du temps pour les anonymes](https://user-images.githubusercontent.com/1523653/112722950-85acb580-8f0c-11eb-8406-3ebf3482e1bd.png)

Pour obtenir l'optimisation, l'astuce a été d'une part de m'assurer que le `select_related` était bien configuré quand on prend le "first message"  car dans la vue on fait un message.topic ce qui générait une requête surnuméraire. 

Ensuite, j'ai profité du fait de savoir que j'étais dans un contexte où je sais que le user est pas authentifié pour simplifier la fonction python qui va chercher le poste

 ## Personnes connectées
![12% de réduction tu temps pour les utilisateurs authentifiés](https://cdn.discordapp.com/attachments/547038766506967041/825365623897849876/unknown.png)

J'ai profité du `select` related pour déjà gagner 5% du temps sur ce chargement. Ensuite, ça a été la chasse aux requêtes dupliquées, surtout à celles que j'étais capable de dédupliquer facilement grâce à la technique du "stockage dans un attribut". C'est le cas pour le "is_read" par exemple même si celui-là en particulier est celui qui a le gain le plus variable en terme de perf.






### Contrôle qualité

- Sans être connecté : aller sur la page d'accueil des forums. Les nombres sont corrects, les URLS sont fonctionnelles
- En étant connecté et en ayant au moins un message non lu qui est le dernier message du forum, aller sur la page d'accueil des forums, la graisse est bien mise, les nombres sont corrects, les urls sont fonctionnelles.

artragis committed:
Enlève des query en double dans la page d'accueil du forum (1/2)

- retire la double sélection du first_post
- met un select_related sur le topic pour éviter une nouvelle requête
- 25% d'amélioration en dev pour les anonymes

artragis committed:
Réduction du nombre de requêtes sur le forum 2/2

- pour les utilisateurs connectés
- 12% de réduction du temps de processing en dev
- quelques ajouts dynamiques de propriétés ajoutées pour faciliter la vie

Situphen reviewed (commented):
Je ne suis pas trop fan de la duplication de `forums.part.html` et `forums.authenticated.part.html`. Je pense qu'on pourrait avoir seulement `forums.part.html` avec un paramètre booléen `is_authenticated` : 

```html+django
        {% if user.is_authenticated %}
            {% include "forum/includes/forums.part.html" is_authenticated=True %}
        {% else %}
            {% include "forum/includes/forums.part.html" is_authenticated=False %}
        {% endif %}
```

et ceci dans `forums.part.html` : 

```html+django
{% if is_authenticated %}
    {% set last_message_url as last_message.topic.resolve_last_read_post_absolute_url %}
{% else %}
    {% set last_message_url as last_message.topic.resolve_first_post_absolute_url %}
{% endif %}
```

puis utiliser `{{ last_message_url }}`

Si `{% set %}` ne fonctionne pas alors utiliser `{% captureas %}` (et dans les deux cas il faut le charger au début du gabarit je crois).

Situphen commented on zds/forum/managers.py:
```diff
@@ -148,6 +150,27 @@ def get_all_messages_of_a_user(self, current, target):
 
 
 class TopicReadManager(models.Manager):
+    def _is_read(self, topic, user=None):
+        """
+        Checks if the user has read the **last post** of the topic.
+        Returns false if the user read the topic except its last post.
+        Technically this is done by checking if the user has a `TopicRead` object
+        for the last post of this topic.
+        :param topic: A topic
+        :param user: A user. If undefined, the current user is used.
+        :return:
+        """
+        if user is None:
+            user = get_current_user()
+        if not user.is_authenticated:
+            return False
+        return self.filter(post=topic.last_message, topic=topic, user=user).exists()
+
+    def topic_read(self, topic, user=None):
+        if not hasattr(topic, "_is_read"):
+            setattr(topic, "_is_read", self._is_read(topic, user))
+        return getattr(topic, "_is_read")
```
Je me demande si ça ne serait pas plus clair en une seule fonction : 

```suggestion
    def topic_read(self, topic, user=None):
        """
        Checks if the user has read the **last post** of the topic.
        Returns false if the user read the topic except its last post.
        Technically this is done by checking if the user has a `TopicRead` object
        for the last post of this topic.
        :param topic: A topic
        :param user: A user. If undefined, the current user is used.
        :return:
        """

        if hasattr(topic, "_is_read"):
            return getattr(topic, "_is_read")

        if user is None:
            user = get_current_user()
        if not user.is_authenticated:
            return False

        is_read = self.filter(post=topic.last_message, topic=topic, user=user).exists()
        setattr(topic, "_is_read", is_read)
        return is_read
```

Situphen replied:
Il faudrait peut-être aussi trouver un nom plus claire que `topic_read()`. Peut-être `is_topic_last_message_read()` ?

artragis replied:
fixed

artragis commented:
Pour la "duplication", au départ, je voulais dupliquer la boucle complète pour justement éviter d'avoir un if/else if dans le template (c'est long et ça peut générer des requêtes inutiles selon ce qu'on teste) mais vue l'organisation, c'était plus compliqué que prévu.

Personnellement, je trouve au contraire que cette "duplication" est nécessaire car on fait beaucoup de choses très différentes en fonction du type de user et que finalement là, on est plus souple.

Situphen commented:
Dans le cas présent on a seulement un appel à une fonction qui change en fonction de si c'est un membre connecté ou un anonyme et le if/else c'est simplement un booléen, donc à mon avis les performances ne risquent pas grand chose alors que dupliquer le code veut dire faire les modifications futures aux deux endroits.

artragis committed:
Fin des optimisations

J'ai trouvé un truc assez énorme quand on fait pas gaffe avec les select_related:
Quand on select une référence vers l'objet lui-même ça empêche
l'edition par la suite
e.g : sélectionner le first_message d'un topic en faisant
un select_related("topic") alors qu'on est en train
d'éditer le topic.

artragis committed:
Opti pour les requêtes issus de back-ref dans les templates

philippemilink reviewed (commented):
Rapport de QA:
----------------

Tout fonctionne bien, mais j'ai juste un doute sur le cas quand le dernier message est non-lu: le lien vers le dernier message non-lu pointe vers l'avant-dernier message du message (celui qui est déjà lu). C'est normal ou bien c'est un bug ?

J'ai relevé les chiffres donnés par la debug toolbar, pour la page `/forums`:
```
non connecté:
default 21.20 ms (73 requêtes including 72 similar and 22 duplicates )
==> default 14.78 ms (40 requêtes including 39 similar )

connecté:
default 49.08 ms (186 requêtes including 174 similar and 101 duplicates )
==> default 66.22 ms (164 requêtes including 143 similar and 37 duplicates )
```
(les temps ne doivent pas être significatifs), oui il y a une réduction, mais ça me paraît encore beaucoup, on ne peut pas faire mieux ?

@Situphen tes remarques ont bien été prises en compte ?

Situphen commented on zds/forum/models.py:
```diff
@@ -130,7 +130,9 @@ def get_last_message(self):
         """
         try:
-            return Post.objects.filter(topic__forum=self).order_by("-pubdate").all()[0]
+            last_post = Post.objects.select_related("topic").filter(topic__forum=self).order_by("-pubdate").all()[0]
```
Tu n'y a pas touché donc tu n'as pas d'obligation à modifier cela mais j'imagine que ça doit simplifier la requête SQL donc vu que c'est appelé plein de fois ce serait chouette de modifier ça !
```suggestion
            last_post = Post.objects.select_related("topic").filter(topic__forum=self).order_by("-pubdate").first()
```

Situphen commented on templates/forum/includes/forums.part.html:
```diff
@@ -25,23 +26,33 @@ <h4 class="topic-title" itemprop="itemListElement">
-                        <span class="forum-last-message">
-                            <span class="forum-last-message-long">
-                                {{ last_message.pubdate|format_date|capfirst }}
+            {% block last_message %}
```
Ce bloc est devenu inutile :)

artragis replied:
fixed

Situphen reviewed (changes requested):
> tes remarques ont bien été prises en compte ?

Oui ! Par contre en regardant le code j'ai une nouvelle remarque.

> Tout fonctionne bien, mais j'ai juste un doute sur le cas quand le dernier message est non-lu: le lien vers le dernier message non-lu pointe vers l'avant-dernier message du message (celui qui est déjà lu). C'est normal ou bien c'est un bug ?

C'est le comportement actuellement en production et c'est normal, c'est pour relire le message précédent et avoir le contexte. :)

> (les temps ne doivent pas être significatifs), oui il y a une réduction, mais ça me paraît encore beaucoup, on ne peut pas faire mieux ?

On peut probablement faire mieux mais ça demanderait encore plus de travail donc autant fusionner le travail qui est fait et on verra plus tard si quelqu'un a la motivation pour améliorer encore plus le code ! :)

artragis commented:
@philippemilink un truc que j'ai remarqué en testant : la première fois il y a beaucoup de requêtes et dès le premier F5, ça change du tout au tout. Il y a probablement un cache quelque part mais je n'arrive pas à comprendre où car la conf de dev ajoute le "NoCacheMiddleware".

artragis committed:
enlève un block inutile

Situphen commented:
> Il y a probablement un cache quelque part mais je n'arrive pas à comprendre où car la conf de dev ajoute le "NoCacheMiddleware".

Ce *middleware* là c'est seulement pour les fichiers statiques :)

Situphen committed:
Merge branch 'dev' into reduce_query_forum
